File: nomikai/ios/NomikaiApp/DB/Records/UserRecord.swift
import Foundation
import GRDB
import Combine

// ─────────────────────────────────────────────────────────────────────────
//  UserRecord - ローカルユーザープロフィール
//
//  使い方:
//    // 作成・保存
//    try UserRecord(id: uuid, name: "太郎", fcmToken: token).save()
//
//    // カレントユーザー取得
//    let me = try UserRecord.findCurrent()
//
//    // FCMトークン更新
//    try me?.updateFcmToken(token)
//
//    // Combine で監視
//    UserRecord.observeCurrent()
//      .sink { user in ... }
// ─────────────────────────────────────────────────────────────────────────

struct UserRecord: ActiveRecord, Codable, Equatable {
    typealias ID = String

    var id: String
    var name: String
    var fcmToken: String?
    var isCurrent: Int          // 1 = このデバイスのユーザー
    var createdAt: Int64
    var updatedAt: Int64

    // ─── DB カラムマッピング ────────────────────────────────────────────
    enum Columns {
        static let id         = Column(CodingKeys.id)
        static let name       = Column(CodingKeys.name)
        static let fcmToken   = Column(CodingKeys.fcmToken)
        static let isCurrent  = Column(CodingKeys.isCurrent)
        static let createdAt  = Column(CodingKeys.createdAt)
        static let updatedAt  = Column(CodingKeys.updatedAt)
    }

    enum CodingKeys: String, CodingKey {
        case id, name
        case fcmToken  = "fcm_token"
        case isCurrent = "is_current"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    static var databaseTableName: String { "users" }

    // GRDB が主キーを自動で認識できるよう定義
    static var persistenceConflictPolicy = PersistenceConflictPolicy(
        insert: .replace,
        update: .replace
    )

    // ─── イニシャライザ ─────────────────────────────────────────────────
    init(
        id: String = UUID().uuidString,
        name: String,
        fcmToken: String? = nil,
        isCurrent: Int = 0,
        createdAt: Int64 = .now,
        updatedAt: Int64 = .now
    ) {
        self.id        = id
        self.name      = name
        self.fcmToken  = fcmToken
        self.isCurrent = isCurrent
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

// ─── 算出プロパティ ───────────────────────────────────────────────────────
extension UserRecord {
    var isCurrentUser: Bool { isCurrent == 1 }
}

// ─── インスタンスメソッド（ドメイン操作） ─────────────────────────────────
extension UserRecord {
    /// FCMトークンを更新して保存する。
    @discardableResult
    func updateFcmToken(_ token: String?) throws -> UserRecord {
        var updated = self
        updated.fcmToken = token
        updated.updatedAt = .now
        return try updated.save()
    }

    /// 名前を更新して保存する。
    @discardableResult
    func updateName(_ newName: String) throws -> UserRecord {
        var updated = self
        updated.name     = newName
        updated.updatedAt = .now
        return try updated.save()
    }
}

// ─── クラスメソッド（ファインダー） ─────────────────────────────────────
extension UserRecord {
    /// IDでユーザーを検索する。
    static func find(_ id: String) throws -> UserRecord? {
        try DatabaseHolder.shared.db.read { db in
            try UserRecord.fetchOne(db, key: id)
        }
    }

    /// カレントユーザーを取得する。
    static func findCurrent() throws -> UserRecord? {
        try DatabaseHolder.shared.db.read { db in
            try UserRecord
                .filter(Columns.isCurrent == 1)
                .fetchOne(db)
        }
    }

    /// 全ユーザーを取得する。
    static func all() throws -> [UserRecord] {
        try DatabaseHolder.shared.db.read { db in
            try UserRecord.fetchAll(db)
        }
    }

    /// カレントユーザーとして設定する（既存の current フラグをリセット）。
    static func setCurrent(_ record: UserRecord) throws {
        try DatabaseHolder.shared.db.write { db in
            try db.execute(sql: "UPDATE users SET is_current = 0")
            var updated = record
            updated.isCurrent = 1
            updated.updatedAt = .now
            try updated.save(db)
        }
    }

    /// カレントユーザーを Combine Publisher で監視する。
    static func observeCurrent() -> AnyPublisher<UserRecord?, Error> {
        ValueObservation
            .tracking { db in
                try UserRecord
                    .filter(Columns.isCurrent == 1)
                    .fetchOne(db)
            }
            .publisher(in: DatabaseHolder.shared.db, scheduling: .immediate)
            .eraseToAnyPublisher()
    }
}

// ─── Int64 now ヘルパー ──────────────────────────────────────────────────
extension Int64 {
    static var now: Int64 { Int64(Date().timeIntervalSince1970 * 1000) }
}

File: nomikai/android/app/src/main/java/com/example/nomikai/data/db/UserRecord.kt
package com.example.nomikai.data.db

import androidx.room.*
import kotlinx.coroutines.flow.Flow

// ─────────────────────────────────────────────
//  Entity
// ─────────────────────────────────────────────
/**
 * ローカルユーザープロフィールのActiveRecordモデル。
 *
 * ```kotlin
 * // 作成・保存
 * UserRecord(id = uuid, name = "太郎", fcmToken = token).save()
 *
 * // 検索
 * val me = UserRecord.findCurrent()
 * val user = UserRecord.find(id)
 *
 * // FCMトークン更新
 * me?.updateFcmToken(newToken)
 *
 * // 削除
 * me?.delete()
 * ```
 */
@Entity(tableName = "users")
data class UserRecord(
    @PrimaryKey val id: String,
    val name: String,
    @ColumnInfo(name = "fcm_token") val fcmToken: String? = null,
    @ColumnInfo(name = "is_current") val isCurrent: Int = 0, // 1 = このデバイスのユーザー
    @ColumnInfo(name = "created_at") val createdAt: Long = System.currentTimeMillis(),
    @ColumnInfo(name = "updated_at") val updatedAt: Long = System.currentTimeMillis()
) : ActiveRecord() {

    // ─── インスタンスメソッド ───────────────────────

    override suspend fun save(): Long =
        DatabaseHolder.db.userDao().upsert(this)

    override suspend fun delete(): Int =
        DatabaseHolder.db.userDao().delete(this)

    /** FCMトークンを更新して保存する */
    suspend fun updateFcmToken(token: String): UserRecord {
        val updated = copy(fcmToken = token, updatedAt = System.currentTimeMillis())
        updated.save()
        return updated
    }

    /** 名前を更新して保存する */
    suspend fun updateName(newName: String): UserRecord {
        val updated = copy(name = newName, updatedAt = System.currentTimeMillis())
        updated.save()
        return updated
    }

    // ─── クラスメソッド (companion) ────────────────

    companion object {
        private val dao get() = DatabaseHolder.db.userDao()

        /** IDでユーザーを検索する */
        suspend fun find(id: String): UserRecord? = dao.findById(id)

        /** 全ユーザーを取得する */
        suspend fun all(): List<UserRecord> = dao.findAll()

        /** このデバイスのカレントユーザーを取得する */
        suspend fun findCurrent(): UserRecord? = dao.findCurrent()

        /** カレントユーザーをFlowで監視する */
        fun observeCurrent(): Flow<UserRecord?> = dao.observeCurrent()

        /** カレントユーザーとして設定（既存のcurrentフラグをリセット） */
        suspend fun setCurrent(record: UserRecord) {
            dao.clearCurrentFlag()
            dao.upsert(record.copy(isCurrent = 1))
        }
    }
}

// ─────────────────────────────────────────────
//  DAO (非公開 - ActiveRecordの内部実装)
// ─────────────────────────────────────────────
@Dao
interface UserDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun upsert(user: UserRecord): Long

    @Delete
    suspend fun delete(user: UserRecord): Int

    @Query("SELECT * FROM users WHERE id = :id LIMIT 1")
    suspend fun findById(id: String): UserRecord?

    @Query("SELECT * FROM users")
    suspend fun findAll(): List<UserRecord>

    @Query("SELECT * FROM users WHERE is_current = 1 LIMIT 1")
    suspend fun findCurrent(): UserRecord?

    @Query("SELECT * FROM users WHERE is_current = 1 LIMIT 1")
    fun observeCurrent(): Flow<UserRecord?>

    @Query("UPDATE users SET is_current = 0")
    suspend fun clearCurrentFlag()
}
